engine/storage/src/main/java/org/apache/cloudstack/storage/endpoint/DefaultEndPointSelector.java

```diff
@@ -128,9 +131,15 @@ protected EndPoint findEndPointInScope(Scope scope, String sqlBase, Long poolId)
             }
         }
 
+        List<Long> dedicatedHosts = dedicatedResourceDao.listAllHosts();
+
         // TODO: order by rand() is slow if there are lot of hosts
         sbuilder.append(") t where t.value<>'true' or t.value is null");    //Added for exclude cluster's subquery
-        sbuilder.append(" ORDER by rand() limit 1");
+        sbuilder.append(" ORDER by ");
+        for (Long hostId: dedicatedHosts){ // put dedicated hosts at the end of the result set
+            sbuilder.append("field(t.id, '" + hostId +"')," );
+        }
+        sbuilder.append(" rand() limit 1");
```

Comment:
if we randomise and limit to one result, what does it matter what the ordering of the set is?

Reply:
We just don't want dedicated hosts ending up being the one selected.

Reply:
The limit is only applied after the sorting.

Reply:
ok, really not very clear code than. What would the resulting sql of this be?

Reply:
```
select t.id from (
    select h.id, cd.value 
    from host h 
    join storage_pool_host_ref s on h.id = s.host_id  
    join cluster c on c.id=h.cluster_id 
    left join cluster_details cd on c.id=cd.cluster_id 
    and cd.name='cluster.storage.operations.exclude' 
    where h.status = 'Up' 
    and h.type = 'Routing' 
    and h.resource_state = 'Enabled' 
    and s.pool_id = 1
    and h.data_center_id = 1
) t 
where t.value<>'true' 
or t.value is null ORDER BY field(t.id, '2'), field(t.id, '5'), rand()
```

Reply:
so why this complicated way to filter out specific values? It isn't very readible and we can add tham to the where clause as `not t.id in (<hostID>, ...)`

Reply:
We don't want to filter them out, just reorder because of the case of all hosts being dedicated.

Reply:
ok, next question, why two separate field() calls and not a combined one?

Reply:
Yes, I have been thinking about that, I may be able to combine this into 1 field call. - I'll see what I can do.